docs/sample-apps/multimodal-chat/backend/api/routes.py

```diff
@@ -0,0 +1,583 @@
+import atexit
+import os
+import shutil
+import tempfile
+from datetime import datetime
+from pathlib import Path
+
+from fastapi import APIRouter, File, HTTPException, UploadFile
+from fastapi.responses import JSONResponse
+from pydantic import BaseModel
+from settings import ALLOWED_TYPES, logger
+
+import pixeltable as pxt
+from pixeltable.functions import openai
+from pixeltable.functions.huggingface import sentence_transformer
+from pixeltable.functions.video import extract_audio
+from pixeltable.iterators import DocumentSplitter
+from pixeltable.iterators.string import StringSplitter
+
+router = APIRouter()
+
+# Configure upload directory
+TEMP_DIR = tempfile.mkdtemp()
+atexit.register(shutil.rmtree, TEMP_DIR)
+logger.info(f"Temporary directory: {Path(TEMP_DIR).absolute()}")
+
+
+class ChatMessage(BaseModel):
+    message: str
+
+
+# Initialize Pixeltable
+pxt.drop_dir("chatbot", force=True)
+pxt.create_dir("chatbot")
+logger.info("Created Pixeltable directory")
+
+docs_table = pxt.create_table(
+    "chatbot.documents",
+    {
+        "document": pxt.Document,
+        "video": pxt.Video,
+        "audio": pxt.Audio,
+        "question": pxt.String,
+    },
+)
+
+conversations = pxt.create_table(
+    "chatbot.conversations",
+    {"role": pxt.String, "content": pxt.String, "timestamp": pxt.Timestamp},
+)
+
+
+# Create prompt function
+@pxt.udf
+def create_prompt(
+    doc_context: list[dict],
+    video_context: list[dict],
+    audio_context: list[dict],
+    question: str,
+) -> str:
+    context_parts = []
+
+    if doc_context:
+        context_parts.append(
+            "Document Context:\n"
+            + "\n\n".join(
+                item["text"] for item in doc_context if item and "text" in item
+            )
+        )
+
+    if video_context:
+        context_parts.append(
+            "Video Context:\n"
+            + "\n\n".join(
+                item["text"] for item in video_context if item and "text" in item
+            )
+        )
+
+    if audio_context:
+        context_parts.append(
+            "Audio Context:\n"
+            + "\n\n".join(
+                item["text"] for item in audio_context if item and "text" in item
+            )
+        )
+
+    full_context = (
+        "\n\n---\n\n".join(context_parts)
+        if context_parts
+        else "No relevant context found."
+    )
+
+    return f"Context:\n{full_context}\n\nQuestion:\n{question}"
+
+
+# Add computed columns for video processing
+docs_table["audio_extract"] = extract_audio(docs_table.video, format="mp3")
```

Comment:
change all the computed column to use:

**.add_computed_column(name**=......)**)**